### ISSUE TITLE
Using codecPayloadType with addTransceiver()

### ISSUE THREAD

**Issue by Orphis** [w3c/webrtc-pc#2008]
The `codecPayloadType` values come from a sender's `getParameters()` call. It seems to me that it isn't then possible to have a `addTransceiver()` call with `sendEncodings` specifying a codec to use (and possibly a scalability mode later).

To force a single codec to be used, I imagine it's possible to use `setCodecPreferences()` with a single entry, but it would prevent negotiation of other codecs, which is less than ideal.

Would the recommended usage be then to:
- Create a generic transceiver with no forced codec in sendEncodings
- Do the negotiation
- Call getParameters() to retrieve the codecPayloadType values
- Call setParameters() to set the codecPayloadType (and possibly the scalability mode)
- Call replaceTrack() to finally start the transmission with the right settings

**Comment by henbos** [comment]
It seems to me that "codecPayloadType" refers to "parameters.codecs[i].payloadType", i.e. from getParameters(). It is referring to sender parameters that don't exist yet, because the sender has not been created yet. Even if it had, there has been no negotiation yet.

If codec payload types are the same for the entire session, and not different depending on m= section, then we could allow setting a payload type that was retrieved from a _different sender_. The spec says what to do if the "codecPayloadType" is not accepted by negotiation (fallback to default), so it's fine if we end up not using this codec. The API makes sense if we can use it as a preference. But this is bad design. Why can't I say "I want to use VP9"? Why do I have to use the specific codecPayloadType?

Would it make more sense if codecPayloadType is read-only, and to select codec we instead set a "preferredCodec" of type RTCRtpCodecCapability? This is something we can obtain from getCapabilities() without having to do a negotiation.

**Comment by aboba** [comment]
@orphis @henbos Will you prepare the slides for discussion of this issue at TPAC?

**Comment by henbos** [comment]
I've prepared slides for this.

**Comment by aboba** [comment]
@Orphis addTransceiver sets the "simulcast envelope" (e.g. the number of encodings and their order) and it is prohibited to change this later with setParameters. Since the maximum number of encodings (and capabilities such as support for SVC) can vary by codec, it would make sense to specify the codecPayloadType in sendEncodings, if only to ensure that createOffer reflects the number of simulcast streams required with the desired codec.  For example, VP8/VP9/AV1 support temporal scalability and H.264/AVC does not, so that if it is desired to have 3 simulcast streams each with 3 temporal layers, this cannot be achieved with H.264/AVC.  Being able to determine this when addTransceiver is called is important so that action (such as trimming the number of simulcast streams or reporting on an unsupported scalabilityMode) can be taken immediately, rather than waiting until O/A completes and discovering that the encodings requested in addTransceiver() do not make sense.

**Comment by aboba** [comment]
@henbos In ORTC, preferredPayloadType is provided in getCapabilities(), which would help the application to correctly choose (and for the user-agent to interpret) the codecPayloadType provided in sendEncodings.

Codec payload types vary between m-lines because in SDP codec are assigned a unique payload type and payload types can't overlap within a BUNDLE group.  Things are a bit looser in ORTC, since there the only requirement is that it be possible to determine which RtpReceiver an inbound RTP packet is for.

**Comment by aboba** [comment]
@henbos I updated the slides.  Hopefully I understand the proposals well enough so that the updates are just clarifications, but please have a look to make sure.

**Comment by henbos** [comment]
@aboba I think preferredPayloadType only works under the assumption that you are the offerer (and thus control the payload types to be negotiated) or you are talking to another endpoint that happen to be use the same implementation as yours. If you are the answerer, you would use whatever payload types that the other endpoint announced, right?

Unless I'm confused about something, I still think we should express this in terms of capabilities, since that is - as far as I can tell - the only way to fully identify a codec without the use of payloadTypes.

If I'm confused about something we can maybe do the slide together.

**Comment by henbos** [comment]
Actually RTCRtpCodecCapability does not seem to cut it either, the only requires fields are mimeType and clockRate. That does not cover all the special configurations. There's sdpFmtpLine that is 'The "format specific parameters" field from the "a=fmtp" line in the SDP corresponding to the codec, if one exists.' but "if one exists" implies this is only after negotiation. Also part of the "a=fmtp" line is the codec payload type, and per [comment](https://github.com/w3c/webrtc-pc/issues/2008#issuecomment-431421369), codec payload types vary between m= sections. Not sure how this is supposed to work.

**Comment by henbos** [comment]
If payload types are predictable before negotiation I think your approach may be correct. Otherwise what I want:

- getCapabilities() should describe codecs (and codec settings) we support in a way that is agnostic towards payload type.
- addTransceiver()/setParameters() should be able to modify the sender's preferred codec with information obtained from getCapabilities().
- getParameters() may still return the codec payload types, but these should not be used for setting because that would only work after negotiation.

**Comment by aboba** [comment]
@henbos Perhaps we can describe two potential approaches with pros/cons and get feedback from the WG?

* Approach 1: Make `codecPayloadType` read-only.  `sendEncodings` passed to `addTransceiver` apply to any codec.  To discover what encodings an codecs are actually set, call `getParameters` after negotiation is complete. Chances of surprise can be minimized using `setCodecPreferences`. 

* Approach 2: Add `preferredPayloadType` to `RTCRtpCodecCapability` and allow `codecPayloadType` to be set. `sendEncodings` is validated based on the codec specified, which must be in the list of codecs returned by `getCapabilities`.  To discover what encodings are actually set, call `getParameters` after negotiation is complete. If the negotiated codec is not the same as in `codecPayloadType` then the specified encodings are not set.  The chance of this can be minimized using `setCodecPreferences`.

**Comment by henbos** [comment]
New PR needed based on TPAC decision.

**Comment by alvestrand** [comment]
Decision at TPAC was to make codecPayloadType read-only. It can't be used for setting.
